pennylane/templates/subroutines/basis_rotation.py

```diff
@@ -0,0 +1,177 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the template for performing basis transformation defined by a set of fermionic ladder operators.
+"""
+
+import pennylane as qml
+from pennylane import numpy as np
+from pennylane.operation import Operation, AnyWires
+from pennylane.qchem.givens_rotations import givens_decomposition
+
+# pylint: disable-msg=too-many-arguments
+class BasisRotation(Operation):
+    r"""Implement a circuit that provides the unitary that can be used to do an exact single-body basis rotation.
+
+    The `~.BasisRotation` template performs a unitary transformation :math:`U(u)` determined by the single-particle fermionic
+    generators as:
+
+    .. math::
+
+        U(u) = \exp{\left( \sum_{pq} \left[\log u \right]_{pq} (a_p^\dagger a_q - a_q^\dagger a_p) \right)}.
+
+    This :math:`U(u)` is implemented efficiently by performing its Givens decomposition into a sequence of
+    :class:`~.PhaseShift` and :class:`~.SingleExcitation` gates using the construction scheme given by
+    `W. Clements et al. <https://opg.optica.org/optica/fulltext.cfm?uri=optica-3-12-1460&id=355743>`_ (2016).
+
+    Args:
+        wires (Iterable[Any]): wires that the operator acts on
+        unitary_matrix (array): matrix specifying the basis transformation
+        check (bool): test unitarity of the provided `unitary_matrix`
+
+    Raises:
+        ValueError: if the provided matrix is not square.
+
+    .. details::
+
+        **Usage Details**
+
+        The `~.BasisRotation` template can be used to implement the evolution :math:`e^{iH}`, where the Hamiltonian
+        :math:`H = \sum_{pq} V_{pq} a^\dagger_p a_q` and :math:`V` is an :math:`N \times N` Hermitian matrix.
+        In this case, the unitary matrix :math:`u` will be the transformation matrix that diagonalizes :math:`V` such that:
+
+        .. math::
+
+            e^{i \sum_{pq} V_{pq} a^\dagger_p a_q} = U(u)^\dagger \prod_k e^{i\lambda_k \sigma_z^k} U(u),
+
+        where the :math:`\lambda_k` are the eigenvalues of the Hamiltonian :math:`H`.
+
+        >>> V = np.array([[ 0.53672126+0.j        , -0.1126064 -2.41479668j],
+        ...               [-0.1126064 +2.41479668j,  1.48694623+0.j        ]])
+        >>> eigen_vals, eigen_vecs = np.linalg.eigh(V)
+        >>> umat = eigen_vecs.T
+        >>> wires = range(len(umat))
+        >>> dev = qml.device("default.qubit", wires=wires)
+        >>> @qml.qnode(dev)
+        ... def circuit():
+        ...    qml.adjoint(qml.BasisRotation(wires=wires, unitary_matrix=umat))
+        ...    for idx, eigenval in enumerate(eigen_vals):
+        ...        qml.RZ(eigenval, wires=[idx])
+        ...    qml.BasisRotation(wires=wires, unitary_matrix=umat)
+        ...    return qml.state()
```

Comment:
Is this needed? Maybe just a quantum function, not a QNode, does the job?

Reply:
Not sure what you mean here. Do you mean there's no need to display the final quantum state?

Reply:
It is a minor point, if the function does not have a return, the code will still work. See [here](https://docs.pennylane.ai/en/stable/code/api/pennylane.matrix.html#usage-details) for example.